plugins/inputs/turbostat/turbostat.go

```diff
@@ -0,0 +1,166 @@
+//go:generate ../../../tools/readme_config_includer/generator
+//go:build linux && amd64
+
+package turbostat
+
+import (
+	"bufio"
+	"context"
+	_ "embed"
+	"errors"
+	"fmt"
+	"io"
+	"os/exec"
+	"sync"
+	"time"
+
+	"github.com/jpillora/backoff"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/influxdata/telegraf/plugins/inputs/turbostat/parser"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+type Turbostat struct {
+	Command []string        `toml:"command"`
+	Log     telegraf.Logger `toml:"-"`
+
+	cancel context.CancelFunc
+}
+
+func (*Turbostat) Gather(telegraf.Accumulator) error {
+	return nil
+}
+
+func (*Turbostat) SampleConfig() string {
+	return sampleConfig
+}
+
+// Start starts the plugin.
+// Maintains a Turbostat process and parses metrics from stdout.
+func (t *Turbostat) Start(a telegraf.Accumulator) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	t.cancel = cancel
+	go func() {
+		retryDelay := &backoff.Backoff{
+			Min:    1 * time.Second,
+			Max:    30 * time.Second,
+			Factor: 2,
+			Jitter: false,
+		}
+		for {
+			t.Log.Info("Starting Turbostat")
+			startedAt := time.Now()
+			err := t.runTurbostat(ctx, a)
+			t.Log.Errorf("Turbostat encountered an error: %s", err)
+			// Exit the goroutine if the context was cancelled.
+			select {
+			case <-ctx.Done():
+				return
+			default:
+			}
+			// If Turbostat ran for more than 30s, there likely was
+			// no major issue and we reset the backoff to the minimum.
+			if time.Since(startedAt) > 30*time.Second {
+				retryDelay.Reset()
+			}
+			delay := retryDelay.Duration()
+			t.Log.Infof("Restarting Turbostat in %.1fs", delay.Seconds())
+			// Wait until the delay elapses or the context is cancelled.
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(delay):
+			}
+		}
+	}()
```

Comment:
Why do you need to reimplement a retry logic? Telegraf supports retrying (see [spec TSD-006](https://github.com/influxdata/telegraf/blob/master/docs/specs/tsd-006-startup-error-behavior.md)) where the user can specify the behavior...

Furthermore, you are retrying on e.g. a missing command or non-existing binary, leaving behind a dysfunctional Telegraf plugin instance that is retried ad infinitum for no good reason...

Reply:
I had a look at [spec TSD-006](https://github.com/influxdata/telegraf/blob/master/docs/specs/tsd-006-startup-error-behavior.md) but I do not understand what I read. Would you mind explaining further please?

Since we cannot determine whether an error is retriable or not (and whichever heuristics we guess may not hold true in future versions of Turbostat anyways), I do not believe there is a good behaviour.
- If we retry, we might retry a process which has no chance to succeed.
- If we do not retry, we might stop the plugin when restarting Turbostat would have solved the problem.

What I did in the meantime is use internal/process, like the execd plugin does.